File: apps/client/lib/aiomatrix/polls.dart
/// MSC3381 / aiomatrix poll wire helpers (stable + unstable namespaces).
library;

const pollStartUnstable = 'org.matrix.msc3381.poll.start';
const pollStartStable = 'm.poll.start';
const pollResponseUnstable = 'org.matrix.msc3381.poll.response';
const pollResponseStable = 'm.poll.response';
const pollEndUnstable = 'org.matrix.msc3381.poll.end';
const pollEndStable = 'm.poll.end';
const pollLeanContentKey = 'dev.aiomatrix.poll';
const pollTextKey = 'org.matrix.msc1767.text';

const pollStartEventTypes = {pollStartUnstable, pollStartStable};
const pollResponseEventTypes = {pollResponseUnstable, pollResponseStable};
const pollEndEventTypes = {pollEndUnstable, pollEndStable};

class PollAnswerOption {
  const PollAnswerOption({required this.id, required this.text});

  final String id;
  final String text;
}

class PollStartData {
  const PollStartData({
    required this.question,
    required this.answers,
    this.maxSelections = 1,
    this.disclosed = true,
  });

  final String question;
  final List<PollAnswerOption> answers;
  final int maxSelections;
  final bool disclosed;
}

Map<String, dynamic>? _asMap(Object? value) {
  if (value is Map) return Map<String, dynamic>.from(value);
  return null;
}

String? _extensibleText(Object? value) {
  final map = _asMap(value);
  if (map == null) return value is String ? value : null;
  final text = map[pollTextKey] ?? map['body'] ?? map['text'];
  return text is String ? text : null;
}

/// Extract structured start block from poll.start event content.
PollStartData? parsePollStartContent(Map<String, dynamic> content) {
  final block = _asMap(content[pollStartUnstable]) ??
      _asMap(content[pollStartStable]);
  if (block == null) return null;

  final question = _extensibleText(block['question'])?.trim() ?? '';
  if (question.isEmpty) return null;

  final rawAnswers = block['answers'];
  if (rawAnswers is! List || rawAnswers.isEmpty) return null;
  final answers = <PollAnswerOption>[];
  for (var i = 0; i < rawAnswers.length; i++) {
    final item = _asMap(rawAnswers[i]);
    if (item == null) continue;
    final id = (item['id'] as String?)?.trim();
    final text = _extensibleText(item)?.trim() ?? '';
    if (id == null || id.isEmpty || text.isEmpty) continue;
    answers.add(PollAnswerOption(id: id, text: text));
  }
  if (answers.isEmpty) return null;

  final max = block['max_selections'];
  final maxSelections = max is int
      ? max.clamp(1, answers.length)
      : (max is num ? max.toInt().clamp(1, answers.length) : 1);

  final kind = block['kind']?.toString() ?? '';
  final disclosed = !kind.contains('undisclosed');

  return PollStartData(
    question: question,
    answers: answers,
    maxSelections: maxSelections,
    disclosed: disclosed,
  );
}

bool isPollStartType(String type) => pollStartEventTypes.contains(type);
bool isPollResponseType(String type) => pollResponseEventTypes.contains(type);
bool isPollEndType(String type) => pollEndEventTypes.contains(type);

String? pollRelationEventId(Map<String, dynamic> content) {
  final relates = _asMap(content['m.relates_to']);
  if (relates == null) return null;
  if (relates['rel_type'] != 'm.reference') return null;
  return relates['event_id'] as String?;
}

List<String> parsePollResponseAnswers(Map<String, dynamic> content) {
  final unstable = _asMap(content[pollResponseUnstable]);
  if (unstable != null) {
    final answers = unstable['answers'];
    if (answers is List) {
      return answers.whereType<String>().toList();
    }
  }
  final stable = _asMap(content[pollResponseStable]);
  if (stable != null) {
    final answers = stable['answers'];
    if (answers is List) {
      return answers.whereType<String>().toList();
    }
  }
  final selections = content['m.selections'];
  if (selections is List) {
    return selections.whereType<String>().toList();
  }
  return const [];
}

/// Build aiomatrix-compatible poll.start content (unstable type preferred).
Map<String, dynamic> buildPollStartContent({
  required String question,
  required List<String> answers,
  int maxSelections = 1,
  bool leanBody = true,
}) {
  final q = question.trim();
  if (q.isEmpty) {
    throw ArgumentError('poll question must be non-empty');
  }
  final cleaned = answers.map((a) => a.trim()).where((a) => a.isNotEmpty).toList();
  if (cleaned.isEmpty) {
    throw ArgumentError('poll requires at least one answer');
  }
  final structured = cleaned.asMap().entries.map((e) {
    return {
      'id': 'answer${e.key}',
      pollTextKey: e.value,
    };
  }).toList();
  final start = {
    'question': {pollTextKey: q},
    'kind': 'org.matrix.msc3381.poll.disclosed',
    'max_selections': maxSelections.clamp(1, cleaned.length),
    'answers': structured,
  };
  final body = leanBody
      ? q
      : '$q\n${cleaned.asMap().entries.map((e) => '${e.key + 1}. ${e.value}').join('\n')}';
  return {
    pollStartUnstable: start,
    pollStartStable: start,
    'body': body,
    'msgtype': 'm.text',
    if (leanBody)
      pollLeanContentKey: {
        'version': 1,
        'lean': true,
        'question': q,
      },
  };
}

Map<String, dynamic> buildPollResponseContent({
  required String pollEventId,
  required List<String> answerIds,
}) {
  final response = {'answers': answerIds};
  return {
    'm.relates_to': {
      'rel_type': 'm.reference',
      'event_id': pollEventId,
    },
    pollResponseUnstable: response,
    pollResponseStable: response,
  };
}

Map<String, dynamic> buildPollEndContent(String pollEventId) {
  const end = <String, dynamic>{};
  return {
    'm.relates_to': {
      'rel_type': 'm.reference',
      'event_id': pollEventId,
    },
    pollEndUnstable: end,
    pollEndStable: end,
    'body': 'Poll ended',
    'msgtype': 'm.text',
  };
}

class PollVoteTally {
  const PollVoteTally({
    required this.counts,
    required this.mySelections,
    required this.totalVoters,
  });

  final Map<String, int> counts;
  final Set<String> mySelections;
  final int totalVoters;
}

/// Latest response per sender wins; empty answers spoil the vote.
PollVoteTally tallyPollVotes({
  required Iterable<({String senderId, DateTime timestamp, List<String> answers})>
      responses,
  required Set<String> validAnswerIds,
  required int maxSelections,
  String? ownUserId,
}) {
  final bySender = <String, ({DateTime timestamp, List<String> answers})>{};
  for (final response in responses) {
    final previous = bySender[response.senderId];
    if (previous != null && !response.timestamp.isAfter(previous.timestamp)) {
      continue;
    }
    bySender[response.senderId] = (
      timestamp: response.timestamp,
      answers: response.answers,
    );
  }

  final counts = {for (final id in validAnswerIds) id: 0};
  var voters = 0;
  var my = <String>{};

  for (final entry in bySender.entries) {
    final truncated = entry.value.answers
        .where(validAnswerIds.contains)
        .take(maxSelections)
        .toSet()
        .toList();
    if (truncated.isEmpty) continue;
    // Spoiled if any unknown id was present before filter and nothing valid left —
    // already filtered. If original had unknown ids, MSC says spoil entirely:
    final original = entry.value.answers;
    if (original.any((id) => !validAnswerIds.contains(id))) continue;
    voters += 1;
    for (final id in truncated) {
      counts[id] = (counts[id] ?? 0) + 1;
    }
    if (ownUserId != null && entry.key == ownUserId) {
      my = truncated.toSet();
    }
  }

  return PollVoteTally(
    counts: counts,
    mySelections: my,
    totalVoters: voters,
  );
}

File: apps/web/src/matrix/polls.ts
/** MSC3381 poll wire helpers (stable + unstable namespaces). */

export const POLL_START_UNSTABLE = "org.matrix.msc3381.poll.start";
export const POLL_START_STABLE = "m.poll.start";
export const POLL_RESPONSE_UNSTABLE = "org.matrix.msc3381.poll.response";
export const POLL_RESPONSE_STABLE = "m.poll.response";
export const POLL_END_UNSTABLE = "org.matrix.msc3381.poll.end";
export const POLL_END_STABLE = "m.poll.end";
export const POLL_TEXT_KEY = "org.matrix.msc1767.text";

export interface PollAnswerOption {
  id: string;
  text: string;
}

export interface PollStartData {
  question: string;
  answers: PollAnswerOption[];
  maxSelections: number;
  disclosed: boolean;
}

export interface PollVoteTally {
  counts: Record<string, number>;
  mySelections: string[];
  totalVoters: number;
  ended: boolean;
}

function record(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function extensibleText(value: unknown): string | undefined {
  if (typeof value === "string") return value;
  const map = record(value);
  if (!map) return undefined;
  const text = map[POLL_TEXT_KEY] ?? map.body ?? map.text;
  return typeof text === "string" ? text : undefined;
}

export function isPollStartType(type: string): boolean {
  return type === POLL_START_UNSTABLE || type === POLL_START_STABLE;
}

export function isPollResponseType(type: string): boolean {
  return type === POLL_RESPONSE_UNSTABLE || type === POLL_RESPONSE_STABLE;
}

export function isPollEndType(type: string): boolean {
  return type === POLL_END_UNSTABLE || type === POLL_END_STABLE;
}

export function parsePollStartContent(content: Record<string, unknown>): PollStartData | null {
  const block = record(content[POLL_START_UNSTABLE]) ?? record(content[POLL_START_STABLE]);
  if (!block) return null;
  const question = extensibleText(block.question)?.trim() ?? "";
  if (!question) return null;
  const rawAnswers = block.answers;
  if (!Array.isArray(rawAnswers) || rawAnswers.length === 0) return null;
  const answers: PollAnswerOption[] = [];
  for (const item of rawAnswers) {
    const map = record(item);
    if (!map) continue;
    const id = typeof map.id === "string" ? map.id.trim() : "";
    const text = extensibleText(map)?.trim() ?? "";
    if (!id || !text) continue;
    answers.push({ id, text });
  }
  if (answers.length === 0) return null;
  const maxRaw = block.max_selections;
  const maxSelections =
    typeof maxRaw === "number"
      ? Math.max(1, Math.min(answers.length, Math.floor(maxRaw)))
      : 1;
  const kind = String(block.kind ?? "");
  return {
    question,
    answers,
    maxSelections,
    disclosed: !kind.includes("undisclosed"),
  };
}

export function pollRelationEventId(content: Record<string, unknown>): string | undefined {
  const relates = record(content["m.relates_to"]);
  if (!relates || relates.rel_type !== "m.reference") return undefined;
  return typeof relates.event_id === "string" ? relates.event_id : undefined;
}

export function parsePollResponseAnswers(content: Record<string, unknown>): string[] {
  for (const key of [POLL_RESPONSE_UNSTABLE, POLL_RESPONSE_STABLE]) {
    const block = record(content[key]);
    if (block && Array.isArray(block.answers)) {
      return block.answers.filter((id): id is string => typeof id === "string");
    }
  }
  if (Array.isArray(content["m.selections"])) {
    return content["m.selections"].filter((id): id is string => typeof id === "string");
  }
  return [];
}

export function buildPollStartContent(input: {
  question: string;
  answers: string[];
  maxSelections?: number;
}): Record<string, unknown> {
  const question = input.question.trim();
  if (!question) throw new Error("Poll question is required");
  const cleaned = input.answers.map((a) => a.trim()).filter(Boolean);
  if (cleaned.length === 0) throw new Error("Poll requires at least one answer");
  const structured = cleaned.map((text, index) => ({
    id: `answer${index}`,
    [POLL_TEXT_KEY]: text,
  }));
  const start = {
    question: { [POLL_TEXT_KEY]: question },
    kind: "org.matrix.msc3381.poll.disclosed",
    max_selections: Math.max(1, Math.min(cleaned.length, input.maxSelections ?? 1)),
    answers: structured,
  };
  return {
    [POLL_START_UNSTABLE]: start,
    [POLL_START_STABLE]: start,
    body: question,
    msgtype: "m.text",
  };
}

export function buildPollResponseContent(pollEventId: string, answerIds: string[]): Record<string, unknown> {
  const response = { answers: answerIds };
  return {
    "m.relates_to": { rel_type: "m.reference", event_id: pollEventId },
    [POLL_RESPONSE_UNSTABLE]: response,
    [POLL_RESPONSE_STABLE]: response,
  };
}

export function buildPollEndContent(pollEventId: string): Record<string, unknown> {
  const end = {};
  return {
    "m.relates_to": { rel_type: "m.reference", event_id: pollEventId },
    [POLL_END_UNSTABLE]: end,
    [POLL_END_STABLE]: end,
    body: "Poll ended",
    msgtype: "m.text",
  };
}

export function tallyPollVotes(input: {
  responses: Array<{ senderId: string; timestamp: number; answers: string[] }>;
  validAnswerIds: Set<string>;
  maxSelections: number;
  ownUserId?: string;
  ended?: boolean;
}): PollVoteTally {
  const bySender = new Map<string, { timestamp: number; answers: string[] }>();
  for (const response of input.responses) {
    const previous = bySender.get(response.senderId);
    if (previous && response.timestamp <= previous.timestamp) continue;
    bySender.set(response.senderId, {
      timestamp: response.timestamp,
      answers: response.answers,
    });
  }

  const counts: Record<string, number> = {};
  for (const id of input.validAnswerIds) counts[id] = 0;
  let totalVoters = 0;
  let mySelections: string[] = [];

  for (const [senderId, value] of bySender) {
    if (value.answers.some((id) => !input.validAnswerIds.has(id))) continue;
    const truncated = [
      ...new Set(
        value.answers
          .filter((id) => input.validAnswerIds.has(id))
          .slice(0, input.maxSelections),
      ),
    ];
    if (truncated.length === 0) continue;
    totalVoters += 1;
    for (const id of truncated) counts[id] = (counts[id] ?? 0) + 1;
    if (input.ownUserId && senderId === input.ownUserId) mySelections = truncated;
  }

  return {
    counts,
    mySelections,
    totalVoters,
    ended: Boolean(input.ended),
  };
}
